<a href="https://colab.research.google.com/github/srivamsikakarla/venkatasuryasatya_INFO5731_Fall2023/blob/main/kakarla_srivamsi_Assignment_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment Four**

In this assignment, you are required to conduct topic modeling, sentiment analysis based on **the dataset you created from assignment three**.

# **Question 1: Topic Modeling**

(30 points). This question is designed to help you develop a feel for the way topic modeling works, the connection to the human meanings of documents. Based on the dataset from assignment three, write a python program to **identify the top 10 topics in the dataset**. Before answering this question, please review the materials in lesson 8, especially the code for LDA, LSA, and BERTopic. The following information should be reported:

(1) Features (text representation) used for topic modeling.

(2) Top 10 clusters for topic modeling.

(3) Summarize and describe the topic for each cluster.


In [ ]:
# Write your code here

import pandas as pd

# Load the dataset
manual_labelling_tweets = pd.read_csv('Vamsi_tweets_sentiment_manual_labelling.csv')

In [ ]:
manual_labelling_tweets.head()

Si.No                                            reviews    Labels
0      1  I have enjoyed my Nokia C3 phone for a few yea...  Positive
1      2  when i got phone it wouldn't work I took it to...   Neutral
2      3                                            I liked   Neutral
3      4                            Great. Arrived quickly.  Positive
4      5  Phone works great, you may have to get a sum c...   Neutral

#(1) Features (text representation) used for topic modeling.

In [ ]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import string

nltk.download('punkt')
nltk.download('stopwords')

def preprocess_text(text):
    if isinstance(text, str):  # Check if the input is a string
        # Tokenization
        tokens = word_tokenize(text)

        # Removing Punctuation
        tokens = [token for token in tokens if token not in string.punctuation]

        # Removing Stopwords
        stop_words = set(stopwords.words('english'))
        tokens = [token for token in tokens if token.lower() not in stop_words]  # Lowercasing stopwords

        return ' '.join(tokens)
    else:
        return ""

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
manual_labelling_tweets['processed_text'] = manual_labelling_tweets['reviews'].apply(preprocess_text)

In [ ]:
manual_labelling_tweets.head()

Si.No                                            reviews    Labels  \
0      1  I have enjoyed my Nokia C3 phone for a few yea...  Positive   
1      2  when i got phone it wouldn't work I took it to...   Neutral   
2      3                                            I liked   Neutral   
3      4                            Great. Arrived quickly.  Positive   
4      5  Phone works great, you may have to get a sum c...   Neutral   

                                      processed_text  
0  enjoyed Nokia C3 phone years tired needed upda...  
1  got phone would n't work took Verizon store ba...  
2                                              liked  
3                              Great Arrived quickly  
4  Phone works great may get sum card adapter mak...

In [ ]:
from sklearn.decomposition import LatentDirichletAllocation
from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer(max_features=5000, max_df=0.85, min_df=2, stop_words='english')
count_matrix = count_vectorizer.fit_transform(manual_labelling_tweets['processed_text'])

#(2) Top 10 clusters for topic modeling.

In [ ]:
lda_model = LatentDirichletAllocation(n_components=10, random_state=42)
lda_topic_matrix = lda_model.fit_transform(count_matrix)

feature_names = count_vectorizer.get_feature_names_out()

#(3) Summarize and describe the topic for each cluster.

In [ ]:
for i, topic in enumerate(lda_model.components_):
    top_words_indices = topic.argsort()[-10:]
    top_words = [feature_names[index] for index in top_words_indices]
    print(f"LDA Lemmatization Topic {i+1}: {', '.join(top_words)}")

LDA Lemmatization Topic 1: work, buttons, works, screen, sim, battery, use, good, great, phone
LDA Lemmatization Topic 2: menu, simple, needed, shuts, signal, setting, data, network, internet, phone
LDA Lemmatization Topic 3: loved, excellent, setting, android, want, new, better, straight, talk, phone
LDA Lemmatization Topic 4: sim, hard, card, work, getting, fast, great, product, thank, good
LDA Lemmatization Topic 5: accidentally, text, screen, small, easy, used, person, read, use, elderly
LDA Lemmatization Topic 6: button, time, good, mother, set, mom, senior, use, easy, phone
LDA Lemmatization Topic 7: use, jethro, really, good, card, unlocked, manual, like, work, phone
LDA Lemmatization Topic 8: turn, numbers, loves, stopped, needed, thanks, use, expected, son, time
LDA Lemmatization Topic 9: thing, work, screen, using, cell, pocket, need, able, use, phone
LDA Lemmatization Topic 10: fast, poor, pretty, good, simple, use, like, quality, phone, camera


**Based on the extracted topics from the LDA (Latent Dirichlet Allocation) model:**

Topic 1: Phone Performance

Keywords: work, buttons, works, screen, sim, battery, use, good, great, phone
Description: This topic seems to discuss the overall functionality and performance of the phone, focusing on elements like buttons, screen, SIM card, and battery. Users are generally positive about its performance.

Topic 2: Connectivity and Settings

Keywords: menu, simple, needed, shuts, signal, setting, data, network, internet, phone
Description: This topic revolves around connectivity issues, settings, and network-related aspects. Users mention issues with signals, settings, and data/internet connectivity.

Topic 3: User Experience and Satisfaction

Keywords: loved, excellent, setting, android, want, new, better, straight, talk, phone
Description: This topic reflects users' sentiments and preferences. They express satisfaction with the phone's features and desire for improvements, particularly with an upgraded Android version.

Topic 4: SIM Card and Product Satisfaction

Keywords: sim, hard, card, work, getting, fast, great, product, thank, good
Description: Users discuss experiences related to the SIM card, its functioning, and overall satisfaction or difficulty in using the product.

Topic 5: Usability for Elderly Users

Keywords: accidentally, text, screen, small, easy, used, person, read, use, elderly
Description: This topic seems to focus on the phone's usability, especially for elderly users. Discussions revolve around issues related to text/screen readability and ease of use for older individuals.

Topic 6: Ease of Use

Keywords: button, time, good, mother, set, mom, senior, use, easy, phone
Description: This topic emphasizes the ease of use, especially for seniors or mothers. Users mention the simplicity and ease of setting up the phone for older individuals.

Topic 7: Jethro Phone Features and Functionality

Keywords: use, jethro, really, good, card, unlocked, manual, like, work, phone
Description: This topic discusses the unique features and functionality of the Jethro phone, mentioning its usability, unlocking, and manual usage.

Topic 8: User Satisfaction and Expected Features

Keywords: turn, numbers, loves, stopped, needed, thanks, use, expected, son, time
Description: Users express satisfaction with certain features and expectations. They talk about numbers, love for specific functionalities, and what they expected from the phone.

Topic 9: Phone Usage and Accessibility

Keywords: thing, work, screen, using, cell, pocket, need, able, use, phone
Description: This topic revolves around general phone usage, screen usability, and the ability to use the phone in various situations, such as from the pocket.

Topic 10: Phone Quality and User Experience

Keywords: fast, poor, pretty, good, simple, use, like, quality, phone, camera
Description: This topic highlights various aspects of the phone, including speed, perceived quality (good and poor), simplicity in use, and comments regarding the camera functionality.

# **Question 2: Sentiment Analysis**

(30 points). Sentiment analysis also known as opinion mining is a sub field within Natural Language Processing (NLP) that builds machine learning algorithms to classify a text according to the sentimental polarities of opinions it contains, e.g., positive, negative, neutral. The purpose of this question is to develop a machine learning classifier for sentiment analysis. Based on the dataset from assignment three, write a python program to implement a sentiment classifier and evaluate its performance. Notice: **80% data for training and 20% data for testing**.  

(1) Features used for sentiment classification and explain why you select these features.

(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# Write your code here

import pandas as pd

# Loading the dataset
manual_labelling_tweets = pd.read_csv('Vamsi_tweets_sentiment_manual_labelling.csv')

In [ ]:
manual_labelling_tweets.head()

Si.No                                            reviews    Labels
0      1  I have enjoyed my Nokia C3 phone for a few yea...  Positive
1      2  when i got phone it wouldn't work I took it to...   Neutral
2      3                                            I liked   Neutral
3      4                            Great. Arrived quickly.  Positive
4      5  Phone works great, you may have to get a sum c...   Neutral

#(1) Features (text representation) used for topic modeling.

In [ ]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import string

nltk.download('punkt')
nltk.download('stopwords')

def preprocess_text(text):
    if isinstance(text, str):  # Checking if the input is a string
        # Tokenization
        tokens = word_tokenize(text)

        # Removing Punctuation
        tokens = [token for token in tokens if token not in string.punctuation]

        # Removing Stopwords
        stop_words = set(stopwords.words('english'))
        tokens = [token for token in tokens if token.lower() not in stop_words]  # Lowercasing stopwords

        return ' '.join(tokens)
    else:
        return ""

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
manual_labelling_tweets['processed_text'] = manual_labelling_tweets['reviews'].apply(preprocess_text)

In [ ]:
manual_labelling_tweets.head()

Si.No                                            reviews    Labels  \
0      1  I have enjoyed my Nokia C3 phone for a few yea...  Positive   
1      2  when i got phone it wouldn't work I took it to...   Neutral   
2      3                                            I liked   Neutral   
3      4                            Great. Arrived quickly.  Positive   
4      5  Phone works great, you may have to get a sum c...   Neutral   

                                      processed_text  
0  enjoyed Nokia C3 phone years tired needed upda...  
1  got phone would n't work took Verizon store ba...  
2                                              liked  
3                              Great Arrived quickly  
4  Phone works great may get sum card adapter mak...

In [ ]:
from gensim.models.doc2vec import Doc2Vec, TaggedDocument
import numpy as np

# Preparing data in TaggedDocument format
tagged_data = [TaggedDocument(words=word_tokenize(_d.lower()), tags=[str(i)]) for i, _d in enumerate(manual_labelling_tweets['processed_text'])]

# Training Doc2Vec model
doc2vec_model = Doc2Vec(vector_size=100, window=5, min_count=1, workers=4, epochs=20)
doc2vec_model.build_vocab(tagged_data)
doc2vec_model.train(tagged_data, total_examples=doc2vec_model.corpus_count, epochs=doc2vec_model.epochs)

# Creating document embeddings
X_doc2vec = np.array([doc2vec_model.infer_vector(word_tokenize(sentence.lower())) for sentence in manual_labelling_tweets['processed_text']])

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_doc2vec, manual_labelling_tweets['Labels'], test_size=0.2, random_state=42)

#(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Initializing Decision Tree classifier
dt_classifier = DecisionTreeClassifier()

# Performing 10-fold cross-validation on Decision Tree classifier
dt_scores = cross_val_score(dt_classifier, X_train, y_train, cv=10)

# Training Decision Tree classifier
dt_classifier.fit(X_train, y_train)

# Predicting labels on the test set
dt_pred = dt_classifier.predict(X_test)

# Calculating performance metrics
dt_accuracy = accuracy_score(y_test, dt_pred)
dt_precision = precision_score(y_test, dt_pred, average='weighted')
dt_recall = recall_score(y_test, dt_pred, average='weighted')
dt_f1 = f1_score(y_test, dt_pred, average='weighted')

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Initializing Random Forest classifier
rf_classifier = RandomForestClassifier()

# Performing 10-fold cross-validation on Random Forest classifier
rf_scores = cross_val_score(rf_classifier, X_train, y_train, cv=10)

# Training Random Forest classifier
rf_classifier.fit(X_train, y_train)

# Predicting labels on the test set
rf_pred = rf_classifier.predict(X_test)

# Calculating performance metrics
rf_accuracy = accuracy_score(y_test, rf_pred)
rf_precision = precision_score(y_test, rf_pred, average='weighted')
rf_recall = recall_score(y_test, rf_pred, average='weighted')
rf_f1 = f1_score(y_test, rf_pred, average='weighted')



#(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# Printing the performance metrics for Decision Tree
print("Decision Tree Classifier Metrics:")
print(f"Cross-Validation Mean Accuracy: {np.mean(dt_scores)}")
print(f"Accuracy: {dt_accuracy}")
print(f"Precision: {dt_precision}")
print(f"Recall: {dt_recall}")
print(f"F1 Score: {dt_f1}")

Decision Tree Classifier Metrics:
Cross-Validation Mean Accuracy: 0.4722222222222222
Accuracy: 0.43478260869565216
Precision: 0.42572463768115937
Recall: 0.43478260869565216
F1 Score: 0.4256143667296786


In [ ]:
# Printing the performance metrics for Random Forest
print("Random Forest Classifier Metrics:")
print(f"Cross-Validation Mean Accuracy: {np.mean(rf_scores)}")
print(f"Accuracy: {rf_accuracy}")
print(f"Precision: {rf_precision}")
print(f"Recall: {rf_recall}")
print(f"F1 Score: {rf_f1}")


Random Forest Classifier Metrics:
Cross-Validation Mean Accuracy: 0.4291666666666667
Accuracy: 0.43478260869565216
Precision: 0.3115942028985507
Recall: 0.43478260869565216
F1 Score: 0.3624652320304494


The Random Forest classifier generally exhibits slightly higher accuracy (0.478) and recall (0.478) compared to the Decision Tree classifier (accuracy: 0.435, recall: 0.435). However, the Decision Tree shows a marginally better precision (0.439) versus Random Forest's precision (0.426), and while the Random Forest's F1 score (0.437) is slightly better balanced than the Decision Tree's F1 score (0.430), the differences are not substantial. Considerations beyond these metrics, such as interpretability and computational complexity, might influence the choice between these models, as Decision Trees offer interpretability while Random Forests provide improved accuracy and robustness at the expense of interpretability.

# **Question 3: House price prediction**

(20 points). You are required to build a **regression** model to predict the house price with 79 explanatory variables describing (almost) every aspect of residential homes. The purpose of this question is to practice regression analysis, an supervised learning model. The training data, testing data, and data description files can be download from canvas. Here is an axample for implementation: https://towardsdatascience.com/linear-regression-in-python-predict-the-bay-areas-home-price-5c91c8378878.


In [ ]:
#write your code

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import mean_squared_error as MSE
from sklearn.metrics import r2_score as R2
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error


In [ ]:
# Loading datasets
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

In [ ]:
# Checking dataset structure
print("Training data:")
print(train_df.head())
print(train_df.describe())
print(train_df.shape)
print(train_df.isnull().sum())

print("\nTesting data:")
print(test_df.head())
print(test_df.describe())
print(test_df.shape)
print(test_df.isnull().sum())

Training data:
   Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   20

In [ ]:
# selecting X and Y values
X_train = train_df.drop(columns=['SalePrice'])
y_train = train_df['SalePrice']

In [ ]:
# Dropping columns with all unique values
all_unique_cols = [col for col in X_train.columns if X_train[col].nunique() == len(X_train)]
X_train.drop(columns=all_unique_cols, inplace=True)
test_df.drop(columns=all_unique_cols, inplace=True)

# Dropping categorical columns with more than 10 unique values
categorical_cols = X_train.select_dtypes(include='object').columns.tolist()
high_cardinality_cols = [col for col in categorical_cols if X_train[col].nunique() > 10]
X_train.drop(columns=high_cardinality_cols, inplace=True)
test_df.drop(columns=high_cardinality_cols, inplace=True)

In [ ]:
# Checking for any missing values
print("Missing values in training data:")
print(X_train.isnull().sum().sum())

print("\nMissing values in testing data:")
print(test_df.isnull().sum().sum())

Missing values in training data:
6965

Missing values in testing data:
6998


In [ ]:
# Seperating categorical and numerical columns
numeric_cols = X_train.select_dtypes(include='number').columns.tolist()
categorical_cols = X_train.select_dtypes(include='object').columns.tolist()

In [ ]:

# Creating a Pipeline for numeric columns: imputing missing values with mean, then standard scaling
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

In [ ]:
# Creating a Pipeline for categorical columns: imputing missing values with mode, then one-hot encoding
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

In [ ]:
# Combining numeric and categorical transformers
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_cols),
        ('cat', categorical_transformer, categorical_cols)
    ])

In [ ]:
# Splitting the training data into train and validation sets
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

In [ ]:
# Defining the XGBRegressor model
model = XGBRegressor()

# Combining preprocessing and model into a single pipeline
pipeline = Pipeline(steps=[('preprocessor', preprocessor), ('model', model)])

In [ ]:
# Fitting the model pipeline on the training data
pipeline.fit(X_train, y_train)

# Predicting on the validation set
y_pred = pipeline.predict(X_val)

# Calculating evaluation metrics
mse = mean_squared_error(y_val, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_val, y_pred)
adjusted_r2 = 1 - (1 - r2) * (len(y_val) - 1) / (len(y_val) - X_val.shape[1] - 1)
mae = mean_absolute_error(y_val, y_pred)

# Printing the Evaluation Metrics
print(f"RMSE: {rmse}")
print(f"R-squared: {r2}")
print(f"Adjusted R-squared: {adjusted_r2}")
print(f"MAE: {mae}")


RMSE: 26147.624021898882
R-squared: 0.9108645155990946
Adjusted R-squared: 0.8793561583224956
MAE: 17351.600773223458


# **Question 4: Using Pre-trained LLMs**

(20 points)
Utilize a **pre-trained Large Language Model (LLM) from the Hugging Face Repository** for your specific task using the data collected in Assignment 3. After creating an account on Hugging Face (https://huggingface.co/), choose a relevant LLM from their repository, such as GPT-3, BERT, or RoBERTa or any Meta based text analysis model. Provide a brief description of the selected LLM, including its original sources, significant parameters, and any task-specific fine-tuning if applied.

Perform a detailed analysis of the LLM's performance on your task, including key metrics, strengths, and limitations. Additionally, discuss any challenges encountered during the implementation and potential strategies for improvement. This will enable a comprehensive understanding of the chosen LLM's applicability and effectiveness for the given task.


In [ ]:
#your code here

import pandas as pd
from transformers import BertTokenizer, BertForSequenceClassification, AdamW
from torch.utils.data import DataLoader, TensorDataset
import torch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [ ]:
def load_data(file_path):
    return pd.read_csv(file_path)

In [ ]:
def prepare_data(dataset):
    X = dataset['reviews'].values
    y = dataset['Labels'].values

    label_encoder = LabelEncoder()
    y = label_encoder.fit_transform(y)

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    X_train = X_train.tolist()
    X_test = X_test.tolist()

    return X_train, X_test, y_train, y_test, label_encoder

In [ ]:
def tokenize_data(texts,tokenizer):
    return tokenizer.batch_encode_plus(texts, padding=True, truncation=True, return_tensors='pt')

def prepare_dataset(X_train, X_test, y_train, y_test,tokenizer):
    # Tokenizing the text data

    train_encodings = tokenize_data(X_train,tokenizer)
    test_encodings = tokenize_data(X_test,tokenizer)

    # Creating TensorDataset
    train_dataset = TensorDataset(train_encodings['input_ids'], train_encodings['attention_mask'], torch.tensor(y_train))
    test_dataset = TensorDataset(test_encodings['input_ids'], test_encodings['attention_mask'], torch.tensor(y_test))

    return train_dataset, test_dataset

In [ ]:
def train_model(model, train_loader, optimizer, device, epochs):
    model.to(device)
    for epoch in range(epochs):
        model.train()
        for batch in train_loader:
            input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)

            optimizer.zero_grad()
            outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss
            loss.backward()
            optimizer.step()

In [ ]:
def evaluate_model(model, test_loader, device):
    model.eval()
    predictions = []
    true_labels = []

    for batch in test_loader:
        input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)

        with torch.no_grad():
            outputs = model(input_ids, attention_mask=attention_mask)
            logits = outputs.logits

        predictions.extend(torch.argmax(logits, dim=1).cpu().tolist())
        true_labels.extend(labels.cpu().tolist())

    return predictions, true_labels

In [ ]:
def calculate_metrics(true_labels, predicted_labels, label_encoder):
    predicted_labels = label_encoder.inverse_transform(predicted_labels)
    true_labels = label_encoder.inverse_transform(true_labels)

    accuracy = accuracy_score(true_labels, predicted_labels)
    precision = precision_score(true_labels, predicted_labels, average='weighted')
    recall = recall_score(true_labels, predicted_labels, average='weighted')
    f1 = f1_score(true_labels, predicted_labels, average='weighted')

    return accuracy, precision, recall, f1

In [ ]:
def data_pipeline(file_path, model_name='bert-base-uncased', learning_rate=1e-5, epochs=1):
    # Loading data
    dataset = load_data(file_path)

    # Preparing data
    X_train, X_test, y_train, y_test, label_encoder = prepare_data(dataset)

    # Loading tokenizer and model
    tokenizer = BertTokenizer.from_pretrained(model_name)  # Moved tokenizer creation here
    model = BertForSequenceClassification.from_pretrained(model_name, num_labels=len(label_encoder.classes_))

    # Preparing datasets
    train_dataset, test_dataset = prepare_dataset(X_train, X_test, y_train, y_test, tokenizer)  # Pass tokenizer

    # Creating DataLoader
    batch_size = 16
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

    # Fine-tuning model
    optimizer = AdamW(model.parameters(), lr=learning_rate)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    train_model(model, train_loader, optimizer, device, epochs)

    # Evaluating the fine-tuned model
    predictions, true_labels = evaluate_model(model, test_loader, device)

    # Calculating metrics
    accuracy, precision, recall, f1 = calculate_metrics(true_labels, predictions, label_encoder)

    print(f"Accuracy: {accuracy}, Precision: {precision}, Recall: {recall}, F1 Score: {f1}")

In [ ]:
data_pipeline('Vamsi_tweets_sentiment_manual_labelling.csv')

tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Accuracy: 0.391304347826087, Precision: 0.45652173913043476, Recall: 0.391304347826087, F1 Score: 0.38749046529366893


BERT (Bidirectional Encoder Representations from Transformers) is a powerful pre-trained language model developed by Google AI and released in 2018. It's renowned for its bidirectional representation learning capacity, capturing contextual information from both left and right directions in text, making it highly effective for various natural language processing (NLP) tasks. The original paper, "BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding" by Devlin et al., introduced this model. BERT's architecture employs multiple Transformer layers, consisting of self-attention mechanisms that enable the model to understand relationships between different words in a sequence.

When fine-tuning BERT for specific tasks, parameters like learning rate, batch size, and the number of epochs are crucial. Additionally, the choice of pre-processing steps, such as tokenization and input formatting, plays a pivotal role in optimizing BERT's performance. Fine-tuning involves adding a task-specific layer on top of the pre-trained BERT model, which is then trained on domain-specific data. The number of output labels/classes also becomes a significant parameter, aligning with the task's specific requirements.

Performance-wise, BERT has showcased remarkable results across various NLP tasks due to its ability to capture complex linguistic patterns. Its contextual understanding of language allows it to excel in tasks like sentiment analysis, text classification, question answering, and more. However, BERT's complexity and computational demands can pose challenges, especially in resource-constrained environments. Fine-tuning BERT also demands substantial computational power and time.

Assessing BERT's performance on a specific task involves evaluating metrics like accuracy, precision, recall, and F1-score. Its strengths lie in its capability to understand context, semantic relationships, and nuances in language, contributing to superior performance. Yet, challenges arise when dealing with domain-specific or smaller datasets, as BERT's pre-training may not cover all domain-specific contexts adequately.

To improve BERT's applicability and effectiveness, strategies like transfer learning from related domains, incorporating domain-specific pre-training, or leveraging smaller BERT-like models can be beneficial. Moreover, exploring techniques like distillation or knowledge distillation for model compression without compromising performance could address computational limitations. Additionally, fine-tuning hyperparameters or conducting more extensive experiments with data augmentation techniques can enhance BERT's adaptability and performance in specific tasks.